### DISCUSSION TITLE
Interpreting the DyNAM rate model & performance tips

### DISCUSSION THREAD

**Discussion by meakbiyik** [stocnet/goldfish#52] [Q&A] [answered]
I have a small question regarding the interpretation of the rate model: for example, how can we describe the coefficient of the time intercept? Is it in the unit of seconds -given that I use timestamps in the `time` column? So in a way, if the intercept is e.g. 5, then under reference factors, an agent is expected to create an event every 5 seconds (as it models the statistics of Poisson distribution). Is this correct?

Also, I have a little issue: for a ~300 node, ~40.000 event DyNAM fit, even the rate model takes several hours to converge. I am experimenting with "default_c" and "gather_compute" engines, which helps, but I really need a bit more performance to conduct the necessary experiments. Do you have any recommendations on how to improve the training speed?

**Comment by auzaheta** [comment, marked as answer]
Hi @meakbiyik,

I'll try to answer your questions in the text below:

- **Rate model interpretation**: You're on the right path. The exponential rate models the waiting time until an actor creates the next relational event. The parameter vector and the set of real functions evaluating the process state are expressed as a linear combination. An exponential transformation of this linear combination ensures that the transition rate cannot become negative. The interpretation of the coefficient as a Poisson distribution uses the exponential of the coefficients. It gives the expected number of events by a unit of time (as you mention, if the unit is in seconds, it provides the expected number of events in one second). An estimate of the average duration until the next event may be obtained using the inverse (1/exp(\theta_0)). You can find a couple of examples using the intercept and additional coefficients in the vignette "How to start" included in `goldfish` version 1.5.1.
- **Performance tips**: I use the following code:

    ```r
    modchoice <- estimate(modelFormula, model = "DyNAM", subModel = "choice",
                            # preprocessingInit = mod02prep, # if you have the preprocess object before hand
                            estimationInit = list(engine = "gather_compute", 
                                                  maxIterations = 100,
                                                  dampingIncreaseFactor = 4, # default = 2, increase if slow
                                                  dampingDecreaseFactor = 6 # default = 3
                                                   ))
    ```      
    Those arguments control the damping of the step size at each iteration of Newton's method. They only have an effect in the first steps and will decrease by half after each iteration.  In particular, the increase factor is the one that is expected to play a role during the first iterations where it's easier to improve the log-likelihood. You might try to change the values and see the changes in time. A relatively high value might have an opposite effect producing large changes in the coefficients that would create more iterations (in a similar vein to the step size parameter in gradient descendent). 
    
    Another place to gain performance is exploiting the OpenMP functionality that was added to the C++ engines.  It'd be tricky (or at least it is for me, I'm not really an OpenMP experienced user). The goldfish installation would check the OpenMP requirements. In the likely case, it doesn't find OpenMP, it would use a single thread C++ functionality. Assuming you set up the OpenMP, you should check it can use the cores/threads to perform calculations (the server I use needs setting up this in advance). OpenMP would let you parallelize some operations during estimation. I think the `RcppArmadillo` or `Rcpp` documentation about OpenMP might be helpful here.

It would be a good addition to this discussion to know what works in your case that helps with the performance. I'd come back if I had some additional tips or detailed information about it.

**Reply by meakbiyik** [reply]
Thanks a lot! Will keep you posted :) Though it seems the issue is with the model not being very scalable to large node sizes (~300), which is not alleviated much when the damping is adjusted, sadly.
